server/src/main/java/org/elasticsearch/action/support/AllowedExecutors.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.util.set.Sets;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A digraph of the executors that are allowed to block while waiting for a future to be completed on a different threadpool. To avoid
+ * deadlocks there must be no cycles in this digraph.
+ */
+class AllowedExecutors {
+    static class AllowedExecutorsEntry implements Map.Entry<String, Set<String>> {
+        private final String key;
+        private final Set<String> value;
+
+        AllowedExecutorsEntry(String key, Set<String> value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        @Override
+        public String getKey() {
+            return key;
+        }
+
+        @Override
+        public Set<String> getValue() {
+            return value;
+        }
+
+        @Override
+        public Set<String> setValue(Set<String> value) {
+            assert false : "unexpected";
+            return null;
+        }
+    }
+
+    /**
+     * @return a digraph of the executors which are permitted to block each other in production.
+     */
+    static Map<String, Set<String>> getProductionAllowedExecutors() {
+        return getAllowedExecutors(
+            newEntry("ccr" /* does not block */),
+            newEntry("clusterApplierService#updateTask" /* does not block */),
+            newEntry("get" /* does not block */),
+            newEntry("masterService#updateTask" /* does not block */),
+            newEntry("searchable_snapshots_cache_fetch_async" /* does not block */),
+            newEntry("system_read" /* does not block */),
+            newEntry("system_write" /* does not block */),
+            newEntry("transport_worker" /* does not block */),
+            newEntry("write" /* does not block */),
+            newEntry(
+                "file-watcher",
+                /* blocks on futures completed by: */
+                "masterService#updateTask"
```

Comment:
Could this be a record with two fields called `blocksOn` and `completedBy` or similar?

Reply:
Yeah ultimately we need to write down some pairs of strings somewhere, but e.g. `searchable_snapshots_cache_fetch_async` is really `org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots#CACHE_FETCH_ASYNC_THREAD_POOL_NAME` which is in `:x-pack:plugin:searchable-snapshots` so we can't write it here.

The question is more how to get those pairs of strings into this currently-statically-defined map. Something something SPI something something perhaps?

Reply:
Could this not be part of the executor builders that we return? I.e., `allowComplete` and `allowCompleteBy`? That would allow either direction to work I think.

Reply:
Yeah defining the pairs of strings somewhere like that should mostly work, it's just about how to get them into this currently-statically-defined map.

Reply:
I see. I wonder if we can make the map non-static and pass it as a thread local instead (noting that none of that will be used in prod, only with assertions enabled).

Reply:
(in fact, then it could be just the set of allowed threads to complete I think).